sourcing-new/src/main/scala/ch/epfl/bluebrain/nexus/sourcingnew/processor/EventSourceProcessor.scala

```diff
@@ -0,0 +1,331 @@
+            }
+          }
+
+        // The actor will be stopped if it doesn't receive any message
+        // during the given duration
+        definition.stopStrategy.lapsedSinceLastInteraction.foreach { duration =>
+          context.setReceiveTimeout(duration, Idle)
+        }
+
+        active(definition.initialState)
+      }
+    }
+
+  implicit private val scheduler: Scheduler = Scheduler(config.evaluationExecutionContext)
+
+  /**
+    * Runs asynchronously the given command and sends
+    * the result back to the same actor which is stashing messages during this period
+    * @param state the current state
+    * @param cmd the command to run
+    * @param context the context of the current actor
+    * @param dryRun if we run as
+    */
+  private def evaluateCommand(
+      state: State,
+      cmd: Command,
+      context: ActorContext[ProcessorCommand],
+      dryRun: Boolean = false
+  ): Unit = {
+    def tellResult(evaluateResult: EvaluationResult) = {
+      val result = if (dryRun) { DryRunResult(evaluateResult) }
+      else { evaluateResult }
+      IO(context.self ! result)
+    }
+
+    val scope = if (dryRun) "testing" else "evaluating"
+    val eval  = for {
+      _ <- IO.shift(config.evaluationExecutionContext)
+      r <- definition.evaluate(state, cmd)
+      _ <- IO.shift(context.executionContext)
+      _ <- tellResult(r.map { e => EvaluationSuccess(e, definition.next(state, e)) }.valueOr(EvaluationRejection(_)))
+    } yield ()
+    val io    = eval
+      .timeoutTo(
+        config.evaluationMaxDuration, {
+          context.log.error2(s"Timed out while $scope command '{}' on actor '{}'", cmd, id)
+          IO.shift(context.executionContext) >> tellResult(EvaluationCommandTimeout(cmd, config.evaluationMaxDuration))
+        }
+      )
+      .onError {
+        case NonFatal(th) =>
+          context.log.error2(s"Error while $scope command '{}' on actor '{}'", cmd, id)
+          IO.shift(context.executionContext) >> tellResult(EvaluationCommandError(cmd, Option(th.getMessage)))
+      }
+
+    io.runAsyncAndForget
+  }
+}
+
+object EventSourceProcessor {
+
+  implicit class PersistentEventDefinitionOps[State, Command, Event: ClassTag, Rejection](
+      definition: PersistentEventDefinition[State, Command, Event, Rejection]
+  ) {
+
+    import definition._
+
+    private def commandHandler(
+        actorContext: ActorContext[EventSourceCommand]
+    ): (State, EventSourceCommand) => Effect[Event, State] = { (state, command) =>
+      command match {
+        case RequestState(_, replyTo: ActorRef[State])            =>
+          Effect.reply(replyTo)(state)
+        case RequestLastSeqNr(_, replyTo: ActorRef[GetLastSeqNr]) =>
+          Effect.reply(replyTo)(
+            GetLastSeqNr(EventSourcedBehavior.lastSequenceNumber(actorContext))
+          )
+        case Append(_, event: Event)                              => Effect.persist(event)
+      }
+    }
+
+    def stateBehavior(
```

Comment:
I would recommend avoiding defining logic in implicit classes; I would suggest moving these functions and calling them from the implicit class.